### PR TITLE
BE - [PG-001] - Create and update links

### LINKED ISSUE #2: BE - [PG001] - Create and update links
As a PO, I want the logic to be developed to create and update the links, so that I can the data in the database can be recorded and modified.

### PR DESCRIPTION
**Pull request by JuanLuisCapellan1**
/ready

### CODE REVIEW

**Comment by JuanLuisCapellan1** [comment]
# Details

This pull request contains the Users and Links CRUD.  It contains five endpoints, these are:

_Users routes:_
- POST /api/login
- POST /api/register
- PATCH /api/change-password
- POST /api/refresh-token

_LInks routes:_
- GET /api/link
- POST /api/create-link
- PATCH /api/update-link

# Testing


<table>
  <thead>
    <tr>
      <th colspan="2">POST /api/login</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>Description</th>
      <td>Logs the user in and returns both accessToken and refreshToken</td>
    </tr>
    <tr>
      <th>Body</th>
      <td>
Content Type: _application/json_ 
Fields:

- username: string

- password: string

</td>
    </tr>
    <tr>
      <th>Returns</th>
      <td>

Content Type: application/json
{
     "accessToken":  accessToken,
     "refreshToken":  refreshToken
}

</td>
    </tr>
  </tbody>
</table>





<table>
  <thead>
    <tr>
      <th colspan="2">POST /api/register</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>Description</th>
      <td>Create a new user and return a message</td>
    </tr>
    <tr>
      <th>Body</th>
      <td>
Content Type: _application/json_

Fields:
- username: string
- password: string
- secret: string
</td>
    </tr>
    <tr>
      <th>Returns</th>
      <td>

Content Type: application/json

{
     message: string
}

</td>
    </tr>
  </tbody>
</table>





<table>
  <thead>
    <tr>
      <th colspan="2">PATCH /api/change-password</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>Description</th>
      <td>Update field password user</td>
    </tr>
    <tr>
      <th>Headers</th>
      <td>
Authorization: Bearer {accessToken}
</td>
    </tr>
    <tr>
      <th>Body</th>
      <td>
Content Type: _application/json_

Fields:
- *username:* string
- *password:* string
- *secret:* string
</td>
    </tr>
    <tr>
      <th>Returns</th>
      <td>

Content Type: application/json

{
    "message":  string
}

</td>
    </tr>
  </tbody>
</table>





<table>
  <thead>
    <tr>
      <th colspan="2">POST /api/refresh-token</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>Description</th>
      <td>Recieves a refreshToken and returns a new accessToken and refreshToken</td>
    </tr>
    <tr>
      <th>Headers</th>
      <td>
Authorization: Bearer {refreshToken}
</td>
    </tr>
    <tr>
      <th>Returns</th>
      <td>

Content Type: application/json

{
    "accessToken": accessToken,
    "refreshToken": refreshToken
}

</td>
    </tr>
  </tbody>
</table>


<table>
  <thead>
    <tr>
      <th colspan="2">POST /api/create-link</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>Description</th>
      <td>Creates a new link</td>
    </tr>
    <tr>
      <th>Headers</th>
      <td>
Authorization: Bearer {accessToken}
</td>
    </tr>
<tr>
      <th>Body</th>
      <td>
Content Type: _multipart/form-data_

Fields:

- name: string
- title: string
- description: string
- btn_name: string
- url: string
- image: file
- detail_result: string
- contain_result: string
</td>
    </tr>
    <tr>
      <th>Returns</th>
      <td>
Content Type: application/json


{
    "link": {
        "name": "llll",
        "title": "ccccccccc",
        "description": "ssssssssss",
        "btn_name": "click",
        "url": "parambcca.com",
        "image": "https://pablo-web-backend.s3.amazonaws.com/b0df4f90-997f-437d-b1ca-f704da403655.jpeg",
        "visibility": "visible",
        "status": "disable",
        "detail_result": "click aqui para obtener el link",
        "contain_result": "la clave es la siguiente: 99212",
        "_id": "62087599888fe9d8cca19e1c",
        "__v": 0
    }
}
</td>
    </tr>
  </tbody>
</table>


<table>
  <thead>
    <tr>
      <th colspan="2">PATCH /api/update-link</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>Description</th>
      <td>Update any field of Link model</td>
    </tr>
    <tr>
      <th>Headers</th>
      <td>
Authorization: Bearer {accessToken}
</td>
    </tr>
<tr>
      <th>Body</th>
      <td>
Content Type: _multipart/form-data_

Fields:
<strong>only fields that must be required are id, visibility and status</strong> 
- name: string
- title: string
- description: string
- btn_name: string
- url: string
- image: file
- visibility: string (Just 'visible' and 'hidden' visibility are valid for this endpoint)
- status: string (Just 'active' and 'disable' status are valid for this endpoint), only one link can be active
- detail_result: string
- contain_result: string
</td>
    </tr>
    <tr>
      <th>Returns</th>
      <td>
Content Type: application/json


{
    "link": {
        "_id": "62087599888fe9d8cca19e1c",
        "name": "test 1",
        "title": "ccccccccc",
        "description": "ssssssssss",
        "btn_name": "click",
        "url": "parambcca.com",
        "image": "https://pablo-web-backend.s3.amazonaws.com/b0df4f90-997f-437d-b1ca-f704da403655.jpeg",
        "visibility": "visible",
        "status": "disable",
        "detail_result": "click aqui para obtener el link",
        "contain_result": "la clave es la siguiente: 99212",
        "__v": 0
    }
}
</td>
    </tr>
  </tbody>
</table>

<table>
  <thead>
    <tr>
      <th colspan="2">GET /api/link</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>Description</th>
      <td>Returns the active link paginate</td>
    </tr>
    <tr>
      <th>Returns</th>
      <td>

Content Type: application/json
{
    "links": {
        "_id": "620867b9b298241f901f4950",
        "name": "complete",
        "title": "ccccccccc",
        "description": "ssssssssss",
        "btn_name": "click",
        "url": "parambcca.com",
        "image": "https://pablo-web-backend.s3.amazonaws.com/3b5f718d-828a-4550-920b-634850584c3d.jpeg",
        "visibility": "visible",
        "status": "active",
        "detail_result": "click aqui para obtener el link",
        "contain_result": "la clave es la siguiente: 99212",
        "__v": 0
    }
}
</td>
    </tr>
  </tbody>
</table>

File: src/app.js [hunk 1]
```diff
@@ -2,25 +2,119 @@ const config = require('./config');
...
     error: {...err, message: err.message, stack: err.stack},
-  }); 
+  });
 });
 
 app.listen(config.PORT, () =>
   console.log(`Server listening on port ${config.PORT}`),
-);
\ No newline at end of file
+);
```
Comment (KingAzizz) [inline src/app.js:116]: add new line

File: src/database.js [hunk 2]
```diff
@@ -20,4 +20,4 @@ function connect() {
   return mongoose.connection;
 }
 
-module.exports = { connect, };
```
Comment (KingAzizz) [inline src/database.js:23]: add new line

File: src/config.js [hunk 3]
```diff
@@ -3,8 +3,10 @@ require('dotenv').config();
...
   AWS_ACCESS_KEY_ID: process.env.AWS_ACCESS_KEY_ID,
+  AWS_S3_BUCKET: process.env.AWS_S3_BUCKET,
   DB_WITH_URL: process.env.DB_CONN_URL || 'mongodb://localhost/test',
   ADMIN_RESET_SECRET: process.env.ADMIN_RESET_SECRET,
   ACCESS_TOKEN_SECRET: process.env.ACCESS_TOKEN_SECRET,
   REFRESH_TOKEN_SECRET: process.env.REFRESH_TOKEN_SECRET,
-}
\ No newline at end of file
+  AWS_SECRET_ACCESS_KEY: process.env.AWS_SECRET_ACCESS_KEY
+}
```
Comment (KingAzizz) [inline src/config.js:12]: add new line

**Review by KingAzizz — CHANGES_REQUESTED** [review]

**Review by angelabrahamnunez — APPROVED** [review]
Excellent!

**Pull request merged**